pkg/kvstore/doublewrite/backend.go

```diff
@@ -0,0 +1,225 @@
+	crdKey, crdErr := (*d.crdBackend).AllocateID(ctx, id, key)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to allocate identity")
+	}
+	kvStoreKey, kvStoreErr := (*d.kvstoreBackend).AllocateID(ctx, id, key)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to allocate identity")
+	}
+	if d.readFromKVStore {
+		return kvStoreKey, kvStoreErr
+	}
+	return crdKey, crdErr
+}
+
+func (d *doubleWriteBackend) AllocateIDIfLocked(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) (allocator.AllocatorKey, error) {
+	crdKey, crdErr := (*d.crdBackend).AllocateIDIfLocked(ctx, id, key, lock)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to allocate identity with lock")
+	}
+	kvStoreKey, kvStoreErr := (*d.kvstoreBackend).AllocateIDIfLocked(ctx, id, key, lock)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to allocate identity with lock")
+	}
+	if d.readFromKVStore {
+		return kvStoreKey, kvStoreErr
+	}
+	return crdKey, crdErr
+}
+
+func (d *doubleWriteBackend) AcquireReference(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	crdErr := (*d.crdBackend).AcquireReference(ctx, id, key, lock)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to acquire reference with lock")
+	}
+	kvStoreErr := (*d.kvstoreBackend).AcquireReference(ctx, id, key, lock)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to acquire reference with lock")
+	}
+	if d.readFromKVStore {
+		return kvStoreErr
+	}
+	return crdErr
+}
+
+func (d *doubleWriteBackend) RunLocksGC(ctx context.Context, staleKeysPrevRound map[string]kvstore.Value) (map[string]kvstore.Value, error) {
+	// This is a no-op for the CRD backend
+	return (*d.kvstoreBackend).RunLocksGC(ctx, staleKeysPrevRound)
+}
+
+func (d *doubleWriteBackend) RunGC(
+	ctx context.Context,
+	rateLimit *rate.Limiter,
+	staleKeysPrevRound map[string]uint64,
+	minID, maxID idpool.ID,
+) (map[string]uint64, *allocator.GCStats, error) {
+	// This is a no-op for the CRD backend
+	return (*d.kvstoreBackend).RunGC(ctx, rateLimit, staleKeysPrevRound, minID, maxID)
+}
+
+func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
+	crdErr := (*d.crdBackend).UpdateKey(ctx, id, key, true)
```

Comment:
It's a bit subtle but it is actually done on purpose... The idea is that when you first switch from `kvstore` to `doublewrite-readkvstore` mode, you want to automatically create CRD identities for kvstore identities which already exist. Forcing this to `true` here was the only solution I found for that. This will force the local key sync logic to create the CRD identities when looping through them here:
https://github.com/cilium/cilium/blob/e3e571c3c7896e5c46f36cd1b1d5b3f8de1ecdef/pkg/allocator/allocator.go#L885-L892

Reply:
I only added a comment for now but I agree that it's not a trivial problem and it would be great to get another opinion of this.
The only comment I have for now is that the solution to "only force it to be true for the CRD case" would not entirely suffice, because ideally we would want to have the possibility to revert from CRD mode back to KVStore mode as well, so we'd need to support both migration paths (KVStore -> CRD and CRD -> KVStore)

Reply:
Oh I see, I read your initial message too quickly, this makes sense 👍